stable/jenkins/templates/_helpers.tpl

```diff
@@ -101,47 +101,17 @@ jenkins:
       serverUrl: "https://kubernetes.default"
       {{- if .Values.agent.enabled }}
       templates:
-      - containers:
-        - alwaysPullImage: {{ .Values.agent.alwaysPullImage }}
-          {{- if .Values.agent.args }}
-          args: "{{ .Values.agent.args }}"
-          {{- else }}
-          args: "^${computer.jnlpmac} ^${computer.name}"
-          {{- end }}
-          command: {{ .Values.agent.command }}
-          envVars:
-          - containerEnvVar:
-              key: "JENKINS_URL"
-              value: "http://{{ template "jenkins.fullname" . }}.{{ template "jenkins.namespace" . }}.svc.{{.Values.clusterZone}}:{{.Values.master.servicePort}}{{ default "" .Values.master.jenkinsUriPrefix }}"
-          {{- if .Values.agent.imageTag }}
-          image: "{{ .Values.agent.image }}:{{ .Values.agent.imageTag }}"
-          {{- else }}
-          image: "{{ .Values.agent.image }}:{{ .Values.agent.tag }}"
-          {{- end }}
-          name: "{{ .Values.agent.sideContainerName }}"
-          privileged: "{{- if .Values.agent.privileged }}true{{- else }}false{{- end }}"
-          resourceLimitCpu: {{.Values.agent.resources.limits.cpu}}
-          resourceLimitMemory: {{.Values.agent.resources.limits.memory}}
-          resourceRequestCpu: {{.Values.agent.resources.requests.cpu}}
-          resourceRequestMemory: {{.Values.agent.resources.requests.memory}}
-          ttyEnabled: {{ .Values.agent.TTYEnabled }}
-          workingDir: "/home/jenkins"
-        idleMinutes: {{ .Values.agent.idleMinutes }}
-        instanceCap: 2147483647
-        {{- if .Values.agent.imagePullSecretName }}
-        imagePullSecrets:
-        - name: {{ .Values.agent.imagePullSecretName }}
-        {{- end }}
-        label: "{{ .Release.Name }}-{{ .Values.agent.componentName }} {{ .Values.agent.customJenkinsLabels  | join " " }}"
-        name: "{{ .Values.agent.podName }}"
-        nodeUsageMode: "NORMAL"
-        podRetention: {{ .Values.agent.podRetention }}
-        showRawYaml: true
-        serviceAccount: "{{ include "jenkins.serviceAccountAgentName" . }}"
-        slaveConnectTimeoutStr: "{{ .Values.agent.slaveConnectTimeout }}"
-        yaml: |-
-          {{ tpl .Values.agent.yamlTemplate . | nindent 10 | trim }}
-        yamlMergeStrategy: "override"
+      {{- include "jenkins.casc.agent" . | nindent 8 }}
+      {{- if .Values.master.JCasC.additionalAgents }}
+      {{- /* save chart agent as `set` modifies the input dictionary */}}
+      {{- $agent := .Values.agent }}
+      {{- /* additionalAgents (each inherits config from agent) */}}
+      {{- range .Values.master.JCasC.additionalAgents }}
+      {{- include "jenkins.casc.agent" (set $ "Values" (set $.Values "agent" (merge . $agent))) | nindent 8 }}
+      {{- end }}
+      {{- /* restore chart agent */}}
+      {{- $_ := set .Values "agent" $agent }}
+      {{- end }}
```

Comment:
I am impressed by your templating skills. Do you thinks it's best to do the agent inheritence via yaml templating. Isn't it possible to configure it in Jenkins that one agent inherits values from another one. It that's possible then I would prefer it that way as users could decide themselves if they want to inherit values or not.

Reply:
We could also set this value so that all templates inherit values:

> defaultsProviderTemplate - The name of the pod template to use for providing default values. When a value is specified and there is a pod template with a matching name, the template will be used as a parent to all other pod templates.

Reply:
> I am impressed by your templating skills.

It was a lot of trial and error :-)

I tried using `defaultsProviderTemplate` but ran into some gotchas as mentioned in the PR notes. Then I discovered `merge` and went that route as I liked how every agent would get the default values the chart provides. Also, I struggled with how to fit the field into the chart since the chart is so "single" agent. But I've since added the `master.JCasC.additionalAgents` field and could add `master.JCasC.defaultAgent`. So I'll revisit `defaultsProviderTemplate` and attempt to document the gotchas for users.